src/Lib.P2P/SecureCommunication/Secio1.cs

```diff
@@ -0,0 +1,229 @@
+                localExchange.Signature = connection.LocalPeerKey.Sign(ms.ToArray());
+            }
+
+            localExchange.EPublicKey = localEphemeralPublicKey;
+            Serializer.SerializeWithLengthPrefix(stream, localExchange, PrefixStyle.Fixed32BigEndian);
+            await stream.FlushAsync(cancel).ConfigureAwait(false);
+
+            // Receive their Exchange packet.  If nothing, then most likely the
+            // remote has closed the connection because it does not like us.
+            var remoteExchange =
+                Serializer.DeserializeWithLengthPrefix<Secio1Exchange>(stream, PrefixStyle.Fixed32BigEndian);
+            if (remoteExchange == null) throw new Exception("Remote refuses the SECIO exchange.");
+
+            // =============================================================================
+            // step 2.1. Verify -- verify their exchange packet is good.
+            var remotePeerKey = Key.CreatePublicKeyFromIpfs(remoteProposal.PublicKey);
+            using (var ms = new MemoryStream())
+            {
+                Serializer.Serialize(ms, remoteProposal);
+                Serializer.Serialize(ms, localProposal);
+                ms.Write(remoteExchange.EPublicKey, 0, remoteExchange.EPublicKey.Length);
+                remotePeerKey.Verify(ms.ToArray(), remoteExchange.Signature);
+            }
+
+            var remoteEphemeralKey = EphermalKey.CreatePublicKeyFromIpfs(curveName, remoteExchange.EPublicKey);
+
+            // =============================================================================
+            // step 2.2. Keys -- generate keys for mac + encryption
+            var sharedSecret = localEphemeralKey.GenerateSharedSecret(remoteEphemeralKey);
+            StretchedKey.Generate(cipherName, hashName, sharedSecret, out var k1, out var k2);
+            if (order < 0)
+            {
+                var tmp = k1;
+                k1 = k2;
+                k2 = tmp;
+            }
+
+            // =============================================================================
+            // step 2.3. MAC + Cipher -- prepare MAC + cipher
+            var secureStream = new Secio1Stream(stream, cipherName, hashName, k1, k2);
+
+            // =============================================================================
+            // step 3. Finish -- send expected message to verify encryption works (send local nonce)
+
+            // Send thier nonce,
+            await secureStream.WriteAsync(remoteProposal.Nonce, 0, remoteProposal.Nonce.Length, cancel)
+               .ConfigureAwait(false);
+            await secureStream.FlushAsync(cancel).ConfigureAwait(false);
+
+            // Receive our nonce.
+            var verification = new byte[localNonce.Length];
+            await secureStream.ReadExactAsync(verification, 0, verification.Length, cancel);
+            if (!localNonce.SequenceEqual(verification)) throw new Exception($"SECIO verification message failure.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)